src/main/java/org/apache/commons/crypto/cipher/Openssl.java

```diff
@@ -43,9 +49,20 @@
     public static final int ENCRYPT_MODE = 1;
     public static final int DECRYPT_MODE = 0;
 
+    static final int DEFAULT_TAG_LEN = 16;
+
+    private int mode = Openssl.DECRYPT_MODE;
+
+    // buffer for AAD data; if consumed, set as null
+    private byte[] aadBuffer;
+    private int tagBitLen = -1;
+
+    // buffer for storing input in decryption, not used for encryption
+    private ByteArrayOutputStream ibuffer = null;
```

Comment:
Is there any more efficient way instead of buffering all the input data in byte array?
Java document said this:
To avoid implementations having to internally buffer ciphertext, all AAD data must be supplied to GCM/CCM implementations (via the updateAAD methods) before the ciphertext is processed (via the update and doFinal methods).